From rpurdie@rpsys.net Sun Oct 30 09:51:40 2005
Subject: USB: Add pxa27x OHCI PM functions
From: Richard Purdie <rpurdie@rpsys.net>
To: Greg KH <gregkh@suse.de>
Date: Sun, 30 Oct 2005 16:33:52 +0000
Message-Id: <1130690033.8248.125.camel@localhost.localdomain>

Add power management functions for the pxa27x USB OHCI host controller.
This is a totally rewritten version of the patch by Nicolas Pitre and
Todd Poynor which accounts for recent USB changes.

Signed-off-by: Richard Purdie <rpurdie@rpsys.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

 drivers/usb/host/ohci-pxa27x.c |   28 ++++++++++++++++++++++++----
 1 file changed, 24 insertions(+), 4 deletions(-)

--- gregkh-2.6.orig/drivers/usb/host/ohci-pxa27x.c	2005-11-12 14:09:13.000000000 -0800
+++ gregkh-2.6/drivers/usb/host/ohci-pxa27x.c	2005-11-12 14:11:24.000000000 -0800
@@ -316,28 +316,48 @@
 	return 0;
 }
 
+#ifdef	CONFIG_PM
 static int ohci_hcd_pxa27x_drv_suspend(struct platform_device *dev, pm_message_t state)
 {
-//	struct usb_hcd *hcd = platform_get_drvdata(dev);
-	printk("%s: not implemented yet\n", __FUNCTION__);
+	struct ohci_hcd *ohci = hcd_to_ohci(dev_get_drvdata(dev));
+
+	if (time_before(jiffies, ohci->next_statechange))
+		msleep(5);
+	ohci->next_statechange = jiffies;
+
+	pxa27x_stop_hc(dev);
+	ohci_to_hcd(ohci)->state = HC_STATE_SUSPENDED;
+	dev->power.power_state = PMSG_SUSPEND;
 
 	return 0;
 }
 
 static int ohci_hcd_pxa27x_drv_resume(struct platform_device *dev)
 {
-//	struct usb_hcd *hcd = platform_get_drvdata(dev);
-	printk("%s: not implemented yet\n", __FUNCTION__);
+	struct ohci_hcd	*ohci = hcd_to_ohci(dev_get_drvdata(dev));
+	int status;
+
+	if (time_before(jiffies, ohci->next_statechange))
+		msleep(5);
+	ohci->next_statechange = jiffies;
+
+	if ((status = pxa27x_start_hc(dev)) < 0)
+		return status;
 
+	dev->power.power_state = PMSG_ON;
+	usb_hcd_resume_root_hub(dev_get_drvdata(dev));
 	return 0;
 }
+#endif
 
 
 static struct platform_driver ohci_hcd_pxa27x_driver = {
 	.probe		= ohci_hcd_pxa27x_drv_probe,
 	.remove		= ohci_hcd_pxa27x_drv_remove,
+#ifdef CONFIG_PM
 	.suspend	= ohci_hcd_pxa27x_drv_suspend, 
 	.resume		= ohci_hcd_pxa27x_drv_resume,
+#endif
 	.driver		= {
 		.name	= "pxa27x-ohci",
 	},
